File: C++/game_of_stack.cpp
#include <iostream>
#include <stdlib.h>
#include <string.h>

using namespace std;

struct stack_CHUNK{
    stack_CHUNK *FD;
    size_t value;
} *stack_baddr, *stack_pointer; // Pointer to the beg of the list & to the end of the list (margins)

size_t option;
size_t num_of_elements;

// Initiate Global variables
void init(){
    stack_baddr = stack_pointer = nullptr;
    option = num_of_elements = 0;
}

// Because the Stack is LIFO, only at the end can be concatenated a new element
void create_stack_element(int content){
  stack_CHUNK *aux;
  // Yet, no elements in the list
  if(stack_baddr == nullptr){
    stack_baddr = new stack_CHUNK;
    (*stack_baddr).value = content;
    (*stack_baddr).FD = nullptr;
    stack_pointer = stack_baddr;
    num_of_elements++;
  }
  // Concatenate new element to the list
  else{
    aux = new stack_CHUNK;
    (*aux).value = content;
    (*aux).FD = nullptr;
    (*stack_pointer).FD = aux;
    stack_pointer = aux;
    num_of_elements++;
  }
}

// Iterate trough the list
static stack_CHUNK* get_penultimate_stack_element(){
  stack_CHUNK *aux;
  for(aux = stack_baddr; aux != nullptr; aux = (*aux).FD){
    if((*(*aux).FD).FD == nullptr){
      return aux;
    }
  }
}

// Because the Stack is LIFO, only the last element can be removed
static void api_pop(){
  if(stack_pointer == nullptr){
    cout << "[*] The Stack is empty" << endl;
  }
  // If there is just one element in the stack
  else if(stack_pointer == stack_baddr){
    delete stack_pointer;
    stack_pointer = stack_baddr = nullptr;
  }
  else{
    stack_CHUNK *aux = get_penultimate_stack_element();
    (*aux).FD = nullptr;
    delete stack_pointer;
    stack_pointer = aux;
  }
}

void stk_print(){
  // The list is empty
  if(stack_pointer == nullptr){
    cout << endl << "[*] Empty Stack" << endl << endl;
  }
  else{
    stack_CHUNK *aux;
    cout << endl << "=========" << endl;
    cout << "THE STACK" << endl;
    for(aux = stack_baddr; aux != nullptr; aux = (*aux).FD){
      cout << (*aux).value << endl;
    }
    cout << "=========" << endl << endl;
  }
}

// Display the interactive MENU
void interactive_menu(){
    cout << "[1] READ" << endl;
    cout << "[2] DUP"<< endl;
    cout << "[3] ADD"<< endl;
    cout << "[4] MUL"<< endl;
    cout << "[5] WRITE"<< endl;
    cout << "[6] PRINT STACK"<< endl;
    cout << "[7] EXIT"<< endl;

    cout << ">> ";cin >> option;
}

// 1. PUSH <integer> pushes <integer> on the stack;
static void api_push(int content){
    create_stack_element(content);
}

 // 2. DUP pushes a copy of the stack head on the stack;
void stk_dup(){
  if(stack_baddr == nullptr){
    cout << endl << "[*] The Stack is empty." << endl << endl;
  }
  else{
    stack_CHUNK *aux;
    aux = new stack_CHUNK;
    (*aux).value = (*stack_pointer).value;
    (*aux).FD = nullptr;
    (*stack_pointer).FD = aux;
    stack_pointer = aux;
    num_of_elements++;
  }
}

 // 3. ADD pops two elements from the stack and pushes their sum on the stack
void stk_add(){
  if(stack_baddr == nullptr || stack_pointer == stack_baddr){
    cout << endl << "[*] There aren't at least 2 elements on the stack" << endl << endl;
  }
  else{
    size_t sum = 0;
    stack_CHUNK *aux = get_penultimate_stack_element();
    sum += (*aux).value + (*(*aux).FD).value;
    api_pop();
    api_pop();
    cout << endl << "[*] The sum is " << sum << endl << endl;
    create_stack_element(sum);
  }
}

// 4. MUL pops two elements from the stack and pushes their product on the stack
void stk_mul(){
  if(stack_baddr == nullptr || stack_pointer == stack_baddr){
    cout << endl << "[*] There aren't at least 2 elements on the stack"<< endl << endl;
  }
  else{
    size_t product = 0;
    stack_CHUNK *aux = get_penultimate_stack_element();
    product += (*aux).value * (*(*aux).FD).value;
    api_pop();
    api_pop();
    cout << endl << "[*] The product is " << product << endl << endl;
    create_stack_element(product);
  }
}

// 5. READ reads an integer from stdin and pushes this value on the stack
void stk_read(){
  size_t aux;
  cout << endl << "[*] Enter the value to push to stak: ";cin >> aux;
  create_stack_element(aux);
}

// 6. WRITE pops an integer value from the stack and prints it on stdout
void stk_write(){
  if(stack_baddr == nullptr){
    cout << endl << "[*] The Stack is empty" << endl << endl;
  }
  else if(stack_baddr == stack_pointer){
    cout << endl << "[*] The poped value from the stack was " << (*stack_pointer).value << endl << endl;
    api_pop();
  }
  else{
    stack_CHUNK *aux = get_penultimate_stack_element();
    cout << endl << "[*] The poped value from the stack was " << (*(*aux).FD).value << endl << endl;
    api_pop();
  }
}

int main(){
    init();
    size_t aux;
    // Take action
    while(true){
        interactive_menu();
        switch(option){
            case 1:
                stk_read();
                break;
            case 2:
                stk_dup();
                break;
            case 3:
                stk_add();
                break;
            case 4:
                stk_mul();
                break;
            case 5:
                stk_write();
                break;
            case 6:
                stk_print();
                break;
            case 7:
                cout << "[*] Good-Bye!" << endl;
                exit(0);
                break;
            default:
                cout << "Unknown option" << endl;
                break;
        }
        // PRINT STACK
        //print_stack();
    }

}

File: Python/game_of_stack.py
#!/usr/bin/env python2.7

from collections import deque, namedtuple
from prettytable import PrettyTable
import time

stack = deque()
t = PrettyTable(['Address', 'Value'])
#f = open("data_in")
#g = open("data_out", 'w')
#data = f.read()
#data = data.split()
option = counter = 0;
addr = 0x0
READ = 1;DUP = 2;ADD = 3;MUL = 4;WRITE = 5;PRINT = 6;EXIT = 7;

def is_number(s):
    try:
        float(s)
        return True
    except ValueError:
        pass
    try:
        import unicodedata
        unicodedata.numeric(s)
        return True
    except (TypeError, ValueError):
        pass
    return False

def welcome():
    print '=================================='
    print '=    WELCOME TO GAME OF STACK    ='
    print '=================================='

def menu():
    print '= [1]READ'
    print '= [2]DUP'
    print '= [3]ADD'
    print '= [4]MUL'
    print '= [5]WRITE'
    print '= [6]PRINT STACK'
    print '= [7]EXIT'

if __name__ == "__main__":
    #start_time = time.time()
    welcome()

    while True:
        menu()
        option = raw_input(">>")
        #option = data[counter]
        #counter += 1
        if len(option) == 0:
            print '\n= [*]Unknown option, enter on of the above.\n'
            continue
        option = int(option, 10)

        # PUSH <integer> pushes <integer> on the stack;
        if option == READ:
            value = raw_input("\n= [*]Enter a value: ")
            #value = data[counter]
            #counter += 1
            print '\n'
            if not is_number(value ):
                print '\n= [*]Please enter a valid integer\n'
                continue
            else:
                stack.append(int(value, 10))
                addr += 1
                t.add_row([hex(addr), value])

        # DUP pushes a copy of the stack head on the stack;
        elif option == DUP:
                if addr < 1:
                    print '\n[*]Not enough elements on the stack to perform this action\n'
                else:
                    stack.append(stack[addr-1])
                    addr += 1
                    t.add_row([hex(addr), stack[addr-2]])

        # ADD pops two elements from the stack and pushes their sum on the stack
        elif option == ADD:
            if addr < 2:
                print '\n[*]Not enough elements on the stack to perform this action\n'
            else:
                temp_sum = stack.pop()
                temp_sum += stack.pop()
                addr -= 1
                stack.append(temp_sum)
                t.del_row(addr)
                t.del_row(addr-1)
                t.add_row([hex(addr), temp_sum])

        # MUL pops two elements from the stack and pushes their product on the stack
        elif option == MUL:
            if addr < 2:
                print '\n[*]Not enough elements on the stack to perform this action\n'
            else:
                temp_product = stack.pop()
                temp_product *= stack.pop()
                addr -= 1
                stack.append(temp_product)
                t.del_row(addr)
                t.del_row(addr-1)
                t.add_row([hex(addr), temp_product])

        # WRITE pops an integer value from the stack and prints it on stdout
        elif option == WRITE:
            if addr < 1:
                print '\n[*]Not enough elements on the stack to perform this action\n'
            else:
                temp_num = stack.pop()
                addr -= 1
                t.del_row(addr)
                print '\n[*]POPed element',temp_num,'\n'

        # PRINT STACK
        elif option == PRINT:
            print '\n',t,'\n'

        # EXIT
        elif option == EXIT:

            #g.write("--- %s seconds ---" % (time.time() - start_time))
            print '\n= [*]Bye-Bye!\n'
            break

        else:
            print '\n= [*]Unknown option, enter on of the above.\n'
